index.js

```diff
@@ -0,0 +1,170 @@
+
+
+function retrieve(program, callback){
+
+  var scratchSpace = 'retriever-scratch' + Math.random()*1e17;
+  fs.mkdirSync(scratchSpace);
+
+
+  function wrappedCb(err, count){
+    var args = arguments;
+    var self = this;
+
+    fs.remove(scratchSpace, function(e){
+      if(err||e){
+       if(callback) return callback(err||e);
+       throw err||e;
+      }
+
+      if(callback) callback.apply(self, args);
+    });
+  }
+
+
+  var stringMatch = typeof program.match === 'string';
+  var regMatch = typeof program.match === 'object';
+  var restrictedFound = 0;
+  
+  var uploadStream;
+
+  var data;
+
+  try{
+    data = JSON.parse(fs.readFileSync(program.file));
+  }catch(err){
+    return wrappedCb(err);
+  }
+
+  var recordCount = data.length;
+  var processed = 0;
+
+
+  function recordCallback(err){
+    if(err) return wrappedCb(err);
+    if(++processed === recordCount) wrappedCb(null, recordCount);
+  }
+
+
+  if(program.bucket) uploadStream = UploadStream(program.bucket, program.profile) 
+
+
+  data.forEach(function(record){
+
+    //Don't allow to traverse to other folders via data.json
+    if(restrictedReg.test(record.name)){
+      restrictedFound = 1;
+      return wrappedCb(new Error('Invalid record name. Must not contain ".." or "/".'));
+    }
+
+    if(restrictedFound) return;
+
+    //If the record is filtered, remove it from the count 
+    if(stringMatch && record.name.indexOf(program.match) === -1 ||
+       regMatch && !program.match.test(record.name)
+    ){
+       if(--recordCount === processed){
+         return wrappedCb(null, recordCount);
+       }
+       return recordCount;
+    }
+
+    var request = hyperquest(record.url);
+    
+    request.on('error', recordCallback);
+
+    checkHash(request, record.hash, function(hashIsEqual, remoteHash){
+      if(hashIsEqual) return; 
+      request.unpipe();
+      request.emit('error', new Error('The hash from ' + record.name + ' did not match the downloaded file\'s hash.\nRecord hash: ' + record.hash +'\nRemote hash: ' + remoteHash +'\n'));
+    });
+
+    if(zipReg.test(record.url)){
+      request.pipe(unzip.Extract({path: path.join(scratchSpace, record.name)}))
+        .on('close', function(){
+          var unzipped = path.join(scratchSpace, record.name, record.file)
+
+          if(csvReg.test(record.file)){
+            csvToVrt(unzipped, record.sourceSrs, function(err, vrt){
+              handleStream(spawnOgr(vrt), record, recordCallback);
+            });
+          }else{
+            handleStream(spawnOgr(unzipped), record, recordCallback);
+          }
+        });
+    }else{
+      if(csvReg.test(record.file)){
+        var csv = path.join(scratchSpace, record.file);
+        var csvStream = fs.createWriteStream(csv);
+
+        csvStream.on('finish', function(err){
+          if(err) return recordCallback(err);
+
+          csvToVrt(csv, record.sourceSrs, function(err, vrt){
+            if(err) return recordCallback(err);
```

Comment:
Cool.  Yeah, it was the race condition scenario that seemed possible if the whole job _didn't_ stop, but since it does, less of an issue.  Of course, as we add more file, we won't want any given failure to bring the whole thing to a halt.

Sounds like you have a good plan for handling this in the future.  Stand by for mergification...